test/src/core/util/logger.spec.js

```diff
@@ -0,0 +1,134 @@
+        log2FuncMapping[Logger.__LOG_LEVEL.NONE] = null;
+        log2FuncMapping[Logger.__LOG_LEVEL.ALWAYS] = logger.always.bind(logger);
+
+        const numericLoglevels = Object.keys(log2FuncMapping).map(s => parseInt(s, 10));
+
+        for (let i = 0; i < numericLoglevels.length; i++) {
+            const numericLoglevel = numericLoglevels[i];
+
+            if (log2FuncMapping[numericLoglevel] === null) {
+                continue;
+            }
+
+            logger.setLogLevel(numericLoglevel);
+
+            // Call logging function
+            log2FuncMapping[numericLoglevel]('I should be logged');
+
+            expect(logger.__log.calls.mostRecent().args[1]).toBe(numericLoglevel);
+
+            logFunctionSpy.calls.reset();
+
+            if (i < numericLoglevels.length - 1) {
+                // Set to higher loglevel
+                logger.setLogLevel(numericLoglevels[i + 1]);
+                // Call logging function
+                log2FuncMapping[numericLoglevel]('I should not be logged');
+
+                expect(logFunctionSpy).not.toHaveBeenCalled();
+            }
+        }
+    });
+
+    it('should have the correct date format', () => {
+        logger.reset();
+
+        const now = 1617712424808;
+        const date = logger.dateFormat.evaluate('', { now });
+
+        expect(date).toBe('2021-04-06T12:33:44.808Z');
+    });
+
+    it('should have the correct message format', () => {
+        const ctx = Logger.createContext('HELLO');
+
+        logger.setLogLevel(Logger.__LOG_LEVEL.INFO);
+        const now = 1617712424808;
+        logger.info('An info', ctx, now);
+
+        expect(logFunctionSpy.calls.mostRecent().args[0]).toBe('2021-04-06T12:33:44.808Z    INFO [test-logger] : An info HELLO');
+    });
+
+    it('should take a given correlation id', () => {
+        logger.setLogLevel(Logger.__LOG_LEVEL.INFO);
+
+        const now = 1617712424808;
+
+        let evtctx = Logger.createEventContext({
+            cid: '338e8dc0-16ca-398b-194c-950ab6d15bd3'
+        });
+
+        logger.info('An info', evtctx, now);
+
+        expect(logFunctionSpy.calls.mostRecent().args[0]).toBe('2021-04-06T12:33:44.808Z    INFO [test-logger] : An info 338e8dc0-16ca-398b-194c-950ab6d15bd3');
+    });
+
+    it('should generate a valid correlation id if not given', () => {
+        const evtctx = Logger.createEventContext({});
+
+        // Check UUIDv4 format
+        expect(evtctx.data).toMatch(/^[a-z0-9]{8}-[a-z0-9]{4}-[a-z0-9]{4}-[a-z0-9]{4}-[a-z0-9]{12}$/);
+    });
+
+    it('should render an error to console.error', () => {
+        spyOn(console, 'error').and.callFake(() => {});
+
+        logger.setLogFunction(null);
+        const err = new Error('An error');
+
+        logger.error(err.message, null, err);
+
+        expect(console.error.calls.mostRecent().args[0]).toMatch(/An error \[Error: .+\]$/g);
+    });
+
+    it('should render an log message console.log', () => {
```

Comment:
I'm missing a test with an invalid object / undefined as message, because I remember that this will lead to a 'strange' error (input.replace is not a function). Maybe this has been fixed already, but a test for such cases would be nice (to avoid to print some strange (and maybe big) things.

Reply:
I will add this test for sure, since it lead to a platform crash

Reply:
Added new function __serializeMessage to logger, which will take care of trying to serialize a given message/object/number... If it fails, placeholders can be specified when the input is `null` or `undefined` `default = '<empty>'` or when it's an unserializable input `default = '<unserializable>'`